aviatrix/resource_aviatrix_aws_tgw_directconnect.go

```diff
@@ -0,0 +1,156 @@
+	client := meta.(*goaviatrix.Client)
+
+	awsTgwDirectConn := &goaviatrix.AwsTgwDirectConn{
+		TgwName:                  d.Get("tgw_name").(string),
+		DirectConnectAccountName: d.Get("directconnect_account_name").(string),
+		DxGatewayID:              d.Get("dx_gateway_id").(string),
+		SecurityDomainName:       d.Get("security_domain_name").(string),
+		AllowedPrefix:            d.Get("allowed_prefix").(string),
+	}
+
+	err := client.CreateAwsTgwDirectConn(awsTgwDirectConn)
+	if err != nil {
+		return fmt.Errorf("failed to create Aviatrix AWS TGW Direct Connection: %s", err)
+	}
+
+	d.SetId(awsTgwDirectConn.TgwName + "~" + awsTgwDirectConn.DxGatewayID)
+	return resourceAviatrixAwsTgwVpnConnRead(d, meta)
+}
+
+func resourceAviatrixAWSTgwDirectConnectRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*goaviatrix.Client)
+
+	tgwName := d.Get("tgw_name").(string)
+	directConnGwID := d.Get("dx_gateway_id").(string)
+
+	if tgwName == "" || directConnGwID == "" {
+		id := d.Id()
+		log.Printf("[DEBUG] Looks like an import. Import Id is %s", id)
+		if !strings.Contains(id, "~") {
+			log.Printf("[DEBUG] Import Id: %s is invalid", id)
+		}
+		d.Set("tgw_name", strings.Split(id, "~")[0])
+		d.Set("dx_gateway_id", strings.Split(id, "~")[1])
+		d.SetId(id)
+	}
+
+	awsTgwDirectConn := &goaviatrix.AwsTgwDirectConn{
+		TgwName:     d.Get("tgw_name").(string),
+		DxGatewayID: d.Get("dx_gateway_id").(string),
+	}
+
+	directConn, err := client.GetAwsTgwDirectConn(awsTgwDirectConn)
+	if err != nil {
+		if err == goaviatrix.ErrNotFound {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("couldn't find Aviatrix Aws Tgw Direct Connection: %s", err)
+	}
+	log.Printf("[INFO] Found Aviatrix Aws Tgw Direct Connection: %#v", directConn)
+
+	d.Set("tgw_name", directConn.TgwName)
+	d.Set("directconnect_account_name", directConn.DirectConnectAccountName)
+	d.Set("dx_gateway_id", directConn.DxGatewayID)
+	d.Set("security_domain_name", directConn.SecurityDomainName)
+	d.Set("allowed_prefix", directConn.AllowedPrefix)
+	d.SetId(directConn.TgwName + "~" + directConn.DxGatewayID)
+
+	return nil
+}
+
+func resourceAviatrixAWSTgwDirectConnectUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*goaviatrix.Client)
+
+	awsTgwDirectConn := &goaviatrix.AwsTgwDirectConn{
+		TgwName:       d.Get("tgw_name").(string),
+		DxGatewayName: d.Get("dx_gateway_id").(string),
+	}
+
+	d.Partial(true)
+
+	log.Printf("[INFO] Updating Aviatrix Site2Cloud: %#v", awsTgwDirectConn)
+	if ok := d.HasChange("allowed_prefix"); ok {
+		awsTgwDirectConn.AllowedPrefix = d.Get("allowed_prefix").(string)
+		err := client.UpdateDirectConnAllowedPrefix(awsTgwDirectConn)
+		if err != nil {
+			return fmt.Errorf("failed to update Aws Tgw Direct Conn Allowed Prefix: %s", err)
+		}
+		d.SetPartial("allowed_prefix")
+	}
+
+	return resourceAviatrixAWSTgwDirectConnectRead(d, meta)
```

Comment:
Use the defer method

Reply:
We do not need use defer method here because only updating allowed_prefix is allowed. If it fails, nothing changes. No need to do a read.